File: app/src/main/java/com/fanchen/imovie/activity/WebActivity.java
package com.fanchen.imovie.activity;

import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.PixelFormat;
import android.net.Uri;
import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.text.TextUtils;
import android.view.ContextMenu;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.TextView;

import com.fanchen.imovie.IMovieAppliction;
import com.fanchen.imovie.R;
import com.fanchen.imovie.base.BaseToolbarActivity;
import com.fanchen.imovie.dialog.BaseAlertDialog;
import com.fanchen.imovie.dialog.OnButtonClickListener;
import com.fanchen.imovie.jsoup.node.Node;
import com.fanchen.imovie.util.DialogUtil;
import com.fanchen.imovie.util.ShareUtil;
import com.fanchen.imovie.util.SystemUtil;
import com.fanchen.imovie.view.ContextMenuTitleView;
import com.fanchen.imovie.view.webview.SwipeWebView;
import com.fanchen.sniffing.SniffingUICallback;
import com.fanchen.sniffing.SniffingVideo;
import com.fanchen.sniffing.x5.SniffingUtil;
import com.tencent.smtt.export.external.interfaces.JsResult;
import com.tencent.smtt.export.external.interfaces.SslError;
import com.tencent.smtt.export.external.interfaces.SslErrorHandler;
import com.tencent.smtt.sdk.TbsVideo;
import com.tencent.smtt.sdk.WebChromeClient;
import com.tencent.smtt.sdk.WebView;
import com.tencent.smtt.sdk.WebViewClient;
import com.xigua.p2p.P2PManager;
import com.xunlei.XLManager;

import java.util.List;

import butterknife.InjectView;


/**
 * webView自定义浏览器页面
 * Created by fanchen on 2017/8/6.
 */
public class WebActivity extends BaseToolbarActivity implements View.OnClickListener, SniffingUICallback {
    public static final String HTMLFLAG = "<SniffingVideo>SniffingVideo</SniffingVideo>";

    public static final String URL = "url";
    public static final String TITLE = "title";
    @InjectView(R.id.swv_content)
    protected SwipeWebView mWebview;
    @InjectView(R.id.fab_play)
    protected FloatingActionButton mButton;

    private String url;
    private Runnable mRunnable = null;

    /**
     * @param context
     * @param url
     */
    public static void startActivity(Context context, String title, String url) {
        try {
            Intent intent = new Intent(context, WebActivity.class);
            intent.putExtra(URL, url);
            intent.putExtra(TITLE, title);
            context.startActivity(intent);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    /**
     * @param context
     * @param url
     */
    public static void startActivity(Context context, String url) {
        startActivity(context, null, url);
    }

    @Override
    protected int getLayout() {
        return R.layout.activity_web;
    }

    @Override
    public String getActivityTitle() {
        return getString(R.string.app_name);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        getWindow().setFormat(PixelFormat.TRANSLUCENT);
        super.onCreate(savedInstanceState);
    }

    @Override
    protected void setListener() {
        super.setListener();
        mButton.setOnClickListener(this);
    }

    @Override
    protected void initActivity(Bundle savedState, LayoutInflater inflater) {
        super.initActivity(savedState, inflater);
        String extra = getIntent().getStringExtra(URL);
        WebView webView = mWebview.getWebView();
        webView.setWebViewClient(new DefaultWebViewClient());
        webView.setWebChromeClient(webChromeClient);
        registerForContextMenu(webView);
        webView.loadUrl(extra);
    }

    @Override
    public void onCreateContextMenu(ContextMenu menu, View v, ContextMenu.ContextMenuInfo menuInfo) {
        super.onCreateContextMenu(menu, v, menuInfo);
        WebView.HitTestResult result = mWebview.getWebView().getHitTestResult();
        if (result.getType() == WebView.HitTestResult.SRC_ANCHOR_TYPE) {
            url = result.getExtra();
            menu.setHeaderView(new ContextMenuTitleView(this, url));
            menu.add(Menu.NONE, 2, 0, "使用第三方程序打开").setOnMenuItemClickListener(menuItemHandler);
            menu.add(Menu.NONE, 0, 1, "复制网址").setOnMenuItemClickListener(menuItemHandler);
        } else if (result.getType() == WebView.HitTestResult.IMAGE_TYPE || result.getType() == WebView.HitTestResult.SRC_IMAGE_ANCHOR_TYPE) {
            url = result.getExtra();
            menu.setHeaderView(new ContextMenuTitleView(this, url));
            menu.add(Menu.NONE, 0, 0, "复制网址").setOnMenuItemClickListener(menuItemHandler);
            menu.add(Menu.NONE, 1, 1, "保存图片").setOnMenuItemClickListener(menuItemHandler);
            menu.add(Menu.NONE, 3, 2, "使用Google搜索").setOnMenuItemClickListener(menuItemHandler);
        }
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (mWebview != null) {
            mWebview.onResume();
        }
    }

    @Override
    protected void onPause() {
        super.onPause();
        if (mWebview != null) {
            mWebview.onPause();
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_web, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_share:
                ShareUtil.share(this, mWebview.getWebView().getTitle(), mWebview.getWebView().getUrl());
                break;
            case R.id.action_start:
                SystemUtil.startThreeApp(this, mWebview.getWebView().getUrl());
                break;
            case R.id.action_copy:
                SystemUtil.putText2Clipboard(this, mWebview.getWebView().getUrl());
                showToast(getString(R.string.clipboard));
                break;
            default:
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        SniffingUtil.get().releaseAll();
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.fab_play:
                String[] stringArray = getResources().getStringArray(R.array.spinner_luxian);
                DialogUtil.showMaterialListDialog(this, stringArray, onItemClickListener);
                break;
        }
    }

    @Override
    public void onBackPressed() {
        if (mWebview != null && mWebview.getWebView().canGoBack()) {
            mWebview.getWebView().goBack();
            return;
        }
        super.onBackPressed();
    }

    private WebChromeClient webChromeClient = new WebChromeClient() {

        @Override
        public boolean onJsConfirm(WebView webView, String s, String s1, JsResult jsResult) {
            if (s1.contains(HTMLFLAG)) {
                if (mButton != null && mButton.getVisibility() != View.VISIBLE) {
                    Node node = new Node(s1);
                    String video = node.attr("video", "src");
                    String source = node.attr("source", "src");
                    String iframe = node.attr("iframe", "src");
                    if (!video.isEmpty() || !iframe.isEmpty() || !source.isEmpty()) {
                        mButton.setVisibility(View.VISIBLE);
                    }
                }
                jsResult.cancel();
                return true;
            }
            return super.onJsConfirm(webView, s, s1, jsResult);
        }

        @Override
        public void onProgressChanged(WebView view, int newProgress) {
            if (newProgress == 100) {
                mWebview.getProgressBar().setVisibility(View.GONE);
                mWebview.getRefreshView().setRefreshing(false);
            } else {
                if (mWebview.getProgressBar().getVisibility() == View.GONE)
                    mWebview.getProgressBar().setVisibility(View.VISIBLE);
                mWebview.getProgressBar().setProgress(newProgress);
            }
        }

    };

    @Override
    public void onSniffingSuccess(View webView, String webUrl, List<SniffingVideo> videos) {
        if (videos.isEmpty()) return;
        String url = videos.get(0).getUrl();
        TbsVideo.openVideo(WebActivity.this, url);
    }

    @Override
    public void onSniffingError(View webView, String url, int errorCode) {
        showSnackbar("解析视频失败");
    }

    @Override
    public void onSniffingStart(View webView, String url) {
        DialogUtil.showProgressDialog(WebActivity.this, "正在解析视频...");
    }

    @Override
    public void onSniffingFinish(View webView, String url) {
        DialogUtil.closeProgressDialog();
    }


    /**
     * // 设置web页面 // 如果页面中链接，如果希望点击链接继续在当前browser中响应， //
     * 而不是新开Android的系统browser中响应该链接，必须覆盖 webview的WebViewClient对象。
     *
     * @author fanchen
     * //
     */
    private class DefaultWebViewClient extends WebViewClient {

        @Override
        public boolean shouldOverrideUrlLoading(WebView view, String url) {
            Uri localUri = Uri.parse(url);
            String scheme = localUri.getScheme();
            if ("http".equalsIgnoreCase(scheme) || "https".equalsIgnoreCase(scheme)) {
                view.loadUrl(url);
            } else if (!TextUtils.isEmpty(url) && IMovieAppliction.app != null && (P2PManager.isXiguaUrl(url) || XLManager.isXLUrlNoHttp(url))) {
                VideoPlayerActivity.startActivity(WebActivity.this, url);
            } else {
                String title = String.format("网页<%s>想打开本地应用，是否允许？", mWebview.getWebView().getOriginalUrl());
                DialogUtil.showCancelableDialog(WebActivity.this, title, new AppButtonClickListener(url));
            }
            return true;
        }

        @Override
        public void onPageStarted(WebView view, String url, Bitmap favicon) {
            super.onPageStarted(view, url, favicon);
            TextView titleView = getTitleView();
            if (isFinishing() || titleView == null || mWebview == null) return;
            String stringExtra = getIntent().getStringExtra(TITLE);
            if (!TextUtils.isEmpty(stringExtra)) {
                titleView.setText(stringExtra);
            } else {
                String title = mWebview.getWebView().getTitle();
                if (TextUtils.isEmpty(title)) {
                    title = getString(R.string.app_name);
                }
                titleView.setText(title);
            }
            mButton.setVisibility(View.GONE);
            try {
                view.setLayerType(View.LAYER_TYPE_HARDWARE, null);
            } catch (Throwable e) {
                e.printStackTrace();
            }
        }

        // 设置程序的标题为网页的标题
        @Override
        public void onPageFinished(WebView view, String url) {
            super.onPageFinished(view, url);
            TextView titleView = getTitleView();
            if (isFinishing() || titleView == null || mWebview == null) return;
            view.setLayerType(View.LAYER_TYPE_NONE, null);
            String stringExtra = getIntent().getStringExtra(TITLE);
            if (!TextUtils.isEmpty(stringExtra)) {
                titleView.setText(stringExtra);
            } else {
                String title = mWebview.getWebView().getTitle();
                if (TextUtils.isEmpty(title)) {
                    title = getString(R.string.app_name);
                }
                titleView.setText(title);
            }
            mWebview.getWebView().getSettings().setBlockNetworkImage(false);
            mWebview.removeCallbacks(mRunnable);
            mWebview.postDelayed(mRunnable = new DelayedRunnable(), 4000);
        }

        // 当load有ssl层的https页面时，如果这个网站的安全证书在Android无法得到认证，
        // WebView就会变成一个空白页，而并不会像PC浏览器中那样跳出一个风险提示框
        @Override
        public void onReceivedSslError(WebView view, SslErrorHandler handler, SslError error) {
            // 忽略证书的错误继续Load页面内容
            handler.proceed();
        }

        @Override
        public void onScaleChanged(WebView view, float oldScale, float newScale) {
            super.onScaleChanged(view, oldScale, newScale);
            view.requestFocus();
            view.requestFocusFromTouch();
        }

    }

    private MenuItem.OnMenuItemClickListener menuItemHandler = new MenuItem.OnMenuItemClickListener() {

        public boolean onMenuItemClick(MenuItem item) {
            switch (item.getItemId()) {
                case 0:
                    SystemUtil.putText2Clipboard(WebActivity.this, url);
                    showSnackbar("复制成功");
                    break;
                case 1:
                    break;
                case 2:
                    SystemUtil.startThreeApp(WebActivity.this, url);
                    break;
                case 3:
                    mWebview.getWebView().loadUrl("https://www.google.com/searchbyimage?image_url=" + url);
                    break;
            }
            return true;
        }

    };

    private class DelayedRunnable implements Runnable {

        @Override
        public void run() {
            if (mWebview == null) return;
            mWebview.loadUrl("javascript:confirm(document.getElementsByTagName('html')[0].innerHTML + '" + HTMLFLAG + "');");
        }

    }

    private class AppButtonClickListener implements OnButtonClickListener {
        private String inteniUrl;

        public AppButtonClickListener(String inteniUrl) {
            this.inteniUrl = inteniUrl;
        }

        @Override
        public void onButtonClick(BaseAlertDialog<?> dialog, int btn) {
            dialog.dismiss();
            if (btn != RIGHT) return;
            SystemUtil.startThreeApp(WebActivity.this, inteniUrl);
        }

    }

    private AdapterView.OnItemClickListener onItemClickListener = new AdapterView.OnItemClickListener() {

        @Override
        public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
            if (position >= WebPlayerActivity.LUXIANS.length || mWebview == null) return;
            String url = mWebview.getWebView().getUrl();
            String parserUrl = String.format(WebPlayerActivity.LUXIANS[position], url);
            String referer = "http://movie.vr-seesee.com/vip";
            SniffingUtil.get().activity(WebActivity.this).referer(referer).url(parserUrl).callback(WebActivity.this).start();
        }

    };

}

File: app/src/main/java/com/fanchen/imovie/activity/WebActivity.kt
package com.fanchen.imovie.activity

import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.PixelFormat
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.os.SystemClock
import android.text.TextUtils
import android.view.*
import android.widget.AdapterView
import com.fanchen.imovie.IMovieAppliction
import com.fanchen.imovie.R
import com.fanchen.imovie.base.BaseToolbarActivity
import com.fanchen.imovie.dialog.BaseAlertDialog
import com.fanchen.imovie.dialog.OnButtonClickListener
import com.fanchen.imovie.jsoup.node.Node
import com.fanchen.imovie.util.DialogUtil
import com.fanchen.imovie.util.ShareUtil
import com.fanchen.imovie.util.SystemUtil
import com.fanchen.imovie.util.VideoUrlUtil
import com.fanchen.imovie.view.ContextMenuTitleView
import com.tencent.smtt.export.external.interfaces.JsResult
import com.tencent.smtt.export.external.interfaces.SslError
import com.tencent.smtt.export.external.interfaces.SslErrorHandler
import com.tencent.smtt.sdk.TbsVideo
import com.tencent.smtt.sdk.WebChromeClient
import com.tencent.smtt.sdk.WebView
import com.tencent.smtt.sdk.WebViewClient
import com.xigua.p2p.P2PManager
import com.xunlei.XLManager
import kotlinx.android.synthetic.main.activity_web.*

/**
 * webView自定义浏览器页面
 */
class WebActivity : BaseToolbarActivity(), View.OnClickListener, VideoUrlUtil.OnParseWebUrlListener {
    override val activityTitle: String
        get() = getString(R.string.app_name)
    private var url: String? = null

    private val webChromeClient = object : WebChromeClient() {

        override fun onJsConfirm(webView: WebView?, s: String?, s1: String?, jsResult: JsResult?): Boolean {
            val node = Node(s1)
            val video = node.attr("video", "src")
            val source = node.attr("source", "src")
            val iframe = node.attr("iframe", "src")
            if (!TextUtils.isEmpty(video) || !TextUtils.isEmpty(iframe) || !TextUtils.isEmpty(source)) {
                if (fab_play != null && fab_play!!.visibility != View.VISIBLE) {
                    fab_play!!.visibility = View.VISIBLE
                }
            }
            jsResult!!.cancel()
            return true
        }

        override fun onProgressChanged(view: WebView?, newProgress: Int) {
            if (newProgress == 100) {
                swv_content.progressBar.visibility = View.GONE
                swv_content.refreshView.isRefreshing = false
            } else {
                if (swv_content.progressBar.visibility == View.GONE)
                    swv_content.progressBar.visibility = View.VISIBLE
                swv_content.progressBar.progress = newProgress
            }
        }

    }


    /**
     * // 设置web页面 // 如果页面中链接，如果希望点击链接继续在当前browser中响应， //
     * 而不是新开Android的系统browser中响应该链接，必须覆盖 webview的WebViewClient对象。
     *
     * @author fanchen
     */
    private val webViewClient = object : WebViewClient() {

        override fun shouldOverrideUrlLoading(view: WebView?, url: String?): Boolean {
            val localUri = Uri.parse(url)
            val scheme = localUri.scheme
            if (scheme!!.equals("http", ignoreCase = true) || scheme.equals("https", ignoreCase = true)) {
                view!!.loadUrl(url)
            } else if (!TextUtils.isEmpty(url) && IMovieAppliction.app != null && (P2PManager.isXiguaUrl(url) || XLManager.isXLUrlNoHttp(url))) {
                VideoPlayerActivity.startActivity(this@WebActivity, url!!)
            } else {
                val title = String.format("网页<%s>想打开本地应用，是否允许？", swv_content!!.webView.originalUrl)
                DialogUtil.showCancelableDialog(this@WebActivity, title, AppButtonClickListener(url!!))
            }
            return true
        }

        override fun onPageStarted(view: WebView, url: String, favicon: Bitmap) {
            if (isFinishing) return
            val stringExtra = intent.getStringExtra(TITLE)
            if (!TextUtils.isEmpty(stringExtra)) {
                mTitleView.text = stringExtra
            } else {
                var title = swv_content!!.webView.title
                if (TextUtils.isEmpty(title)) {
                    title = getString(R.string.app_name)
                }
                mTitleView.text = title
            }
            fab_play!!.visibility = View.GONE
            try {
                view.setLayerType(View.LAYER_TYPE_HARDWARE, null)
            } catch (e: Throwable) {
                e.printStackTrace()
            }

        }

        // 设置程序的标题为网页的标题
        override fun onPageFinished(view: WebView?, url: String?) {
            if (isFinishing) return
            view!!.setLayerType(View.LAYER_TYPE_NONE, null)
            val stringExtra = intent.getStringExtra(TITLE)
            if (!TextUtils.isEmpty(stringExtra)) {
                mTitleView.text = stringExtra
            } else {
                var title = swv_content!!.webView.title
                if (TextUtils.isEmpty(title)) {
                    title = getString(R.string.app_name)
                }
                mTitleView.text = title
            }
            swv_content!!.webView.settings.blockNetworkImage = false
        }

        // 当load有ssl层的https页面时，如果这个网站的安全证书在Android无法得到认证，
        // WebView就会变成一个空白页，而并不会像PC浏览器中那样跳出一个风险提示框
        override fun onReceivedSslError(view: WebView?, handler: SslErrorHandler, error: SslError?) {
            // 忽略证书的错误继续Load页面内容
            handler.proceed()
        }

        override fun onScaleChanged(view: WebView?, oldScale: Float, newScale: Float) {
            super.onScaleChanged(view, oldScale, newScale)
            view!!.requestFocus()
            view.requestFocusFromTouch()
        }

    }

    private val sniffingRunnable = Runnable {
        while (!isFinishing && swv_content != null && swv_content!!.webView != null) {
            SystemClock.sleep(3000)
            swv_content!!.loadUrl("javascript:confirm(document.getElementsByTagName('html')[0].innerHTML);")
        }
    }

    private val menuItemHandler = MenuItem.OnMenuItemClickListener { item ->
        when (item.itemId) {
            0 -> {
                SystemUtil.putText2Clipboard(this@WebActivity, url)
                showSnackbar("复制成功")
            }
            1 -> {
            }
            2 -> SystemUtil.startThreeApp(this@WebActivity, url)
            3 -> swv_content!!.webView.loadUrl("https://www.google.com/searchbyimage?image_url=" + url!!)
        }
        true
    }

    private val onItemClickListener = AdapterView.OnItemClickListener { _, _, position, _ ->
        if (position >= WebPlayerActivity.LUXIANS.size || swv_content == null) return@OnItemClickListener
        val url = swv_content!!.webView.url
        val parserUrl = String.format(WebPlayerActivity.LUXIANS[position], url)
        DialogUtil.showProgressDialog(this@WebActivity, "正在解析视频...")
        val referer = "http://movie.vr-seesee.com/vip"
        VideoUrlUtil.getInstance().setLoadUrl(parserUrl, referer).startParse()
    }

    override fun getLayout(): Int {
        return R.layout.activity_web
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        window.setFormat(PixelFormat.TRANSLUCENT)
        super.onCreate(savedInstanceState)
    }

    override fun setListener() {
        super.setListener()
        fab_play!!.setOnClickListener(this)
    }

    override fun initActivity(savedState: Bundle?, inflater: LayoutInflater) {
        super.initActivity(savedState, inflater)
        val webView = swv_content!!.webView
        VideoUrlUtil.getInstance().init(this).setParserTime(2 * 1000).setOnParseListener(this).setAutoDestroy(false)
        val webSetting = webView.settings
        webSetting.javaScriptCanOpenWindowsAutomatically = true
        webSetting.javaScriptEnabled = true
        webSetting.setSupportMultipleWindows(false)
        webSetting.databaseEnabled = true
        webSetting.setGeolocationEnabled(true)
        webSetting.domStorageEnabled = true// 开启DOM
        webSetting.allowFileAccess = true// 设置支持文件流
        webSetting.useWideViewPort = true// 调整到适合webview大小
        webSetting.loadWithOverviewMode = true// 调整到适合webview大小
        webSetting.blockNetworkImage = true// 提高网页加载速度，暂时阻塞图片加载，然后网页加载好了，在进行加载图片
        webSetting.setAppCacheEnabled(true)// 开启缓存机制
        webSetting.setAppCacheMaxSize((64 * 1024 * 1024).toLong())
        webSetting.cacheMode = android.webkit.WebSettings.LOAD_DEFAULT
        webSetting.setAppCachePath(getDir("cache", Context.MODE_PRIVATE).path)
        webSetting.setGeolocationDatabasePath(getDir("database", Context.MODE_PRIVATE).path)
        webSetting.userAgentString = " Mozilla/5.0 (iPhone; U; CPU iPhone OS 5_0 like Mac OS X; en-us) AppleWebKit/534.46 (KHTML, like Gecko) Version/5.1 Mobile/9A334 Safari/7534.48.3"
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            webSetting.mixedContentMode = 0
        }
        webView.setWebViewClient(webViewClient)
        webView.setWebChromeClient(webChromeClient)
        registerForContextMenu(webView)
        Thread(sniffingRunnable).start()
        if (intent.hasExtra(URL)) {
            swv_content!!.loadUrl(intent.getStringExtra(URL))
        }
    }

    override fun onCreateContextMenu(menu: ContextMenu, v: View, menuInfo: ContextMenu.ContextMenuInfo) {
        super.onCreateContextMenu(menu, v, menuInfo)
        val result = swv_content!!.webView.hitTestResult
        if (result.type == WebView.HitTestResult.SRC_ANCHOR_TYPE) {
            url = result.extra
            menu.setHeaderView(ContextMenuTitleView(this, url))
            menu.add(Menu.NONE, 2, 0, "使用第三方程序打开").setOnMenuItemClickListener(menuItemHandler)
            menu.add(Menu.NONE, 0, 1, "复制网址").setOnMenuItemClickListener(menuItemHandler)
        } else if (result.type == WebView.HitTestResult.IMAGE_TYPE || result.type == WebView.HitTestResult.SRC_IMAGE_ANCHOR_TYPE) {
            url = result.extra
            menu.setHeaderView(ContextMenuTitleView(this, url))
            menu.add(Menu.NONE, 0, 0, "复制网址").setOnMenuItemClickListener(menuItemHandler)
            menu.add(Menu.NONE, 1, 1, "保存图片").setOnMenuItemClickListener(menuItemHandler)
            menu.add(Menu.NONE, 3, 2, "使用Google搜索").setOnMenuItemClickListener(menuItemHandler)
        }
    }

    override fun onResume() {
        super.onResume()
        if (swv_content != null) {
            swv_content!!.onResume()
        }
    }

    override fun onPause() {
        super.onPause()
        if (swv_content != null) {
            swv_content!!.onPause()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_web, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_share -> ShareUtil.share(this, swv_content!!.webView.title, swv_content!!.webView.url)
            R.id.action_start -> SystemUtil.startThreeApp(this, swv_content!!.webView.url)
            R.id.action_copy -> {
                SystemUtil.putText2Clipboard(this, swv_content!!.webView.url)
                showToast(getString(R.string.clipboard))
            }
            else -> {
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onDestroy() {
        if (swv_content != null && swv_content!!.webView != null) {
            swv_content!!.webView.destroy()
        }
        VideoUrlUtil.getInstance().destroy()
        super.onDestroy()
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.fab_play -> {
                val stringArray = resources.getStringArray(R.array.spinner_luxian)
                DialogUtil.showMaterialListDialog(this, stringArray, onItemClickListener)
            }
        }
    }

    override fun onBackPressed() {
        if (swv_content != null && swv_content!!.webView.canGoBack()) {
            swv_content!!.webView.goBack()
            return
        }
        super.onBackPressed()
    }

    override fun onFindUrl(url: String) {
        TbsVideo.openVideo(this@WebActivity, url)
        DialogUtil.closeProgressDialog()
    }

    override fun onError(errorMsg: String) {
        showToast("视频解析错误")
        DialogUtil.closeProgressDialog()
    }

    private inner class AppButtonClickListener(private val inteniUrl: String) : OnButtonClickListener {

        override fun onButtonClick(dialog: BaseAlertDialog<*>, btn: Int) {
            dialog.dismiss()
            if (btn != OnButtonClickListener.RIGHT) return
            SystemUtil.startThreeApp(this@WebActivity, inteniUrl)
        }

    }

    companion object {

        const val URL = "url"
        const val TITLE = "title"

        fun startActivity(context: Context, title: String?, url: String) {
            try {
                val intent = Intent(context, WebActivity::class.java)
                intent.putExtra(URL, url)
                intent.putExtra(TITLE, title)
                context.startActivity(intent)
            } catch (e: Exception) {
                e.printStackTrace()
            }

        }

        fun startActivity(context: Context, url: String) {
            startActivity(context, null, url)
        }
    }
}
